pipeline/dags/parallel_download.py

```diff
@@ -0,0 +1,198 @@
+
+from airflow import DAG
+from airflow.operators.empty import EmptyOperator
+from airflow.providers.http.sensors.http import HttpSensor
+from airflow.providers.ssh.hooks.ssh import SSHHook
+from airflow.hooks.base import BaseHook
+from airflow.decorators import task
+from airflow.utils.task_group import TaskGroup
+
+from harvester.pmh_interface import PMH_API
+from harvester import utils
+
+from operators.custom_operators import (
+    SaveMetsSFTPOperator,
+    SaveAltosForMetsSFTPOperator,
+    CreateConnectionOperator,
+)
+
+HTTPConnection.debuglevel = 1
+
+logging.basicConfig(filename="/home/ubuntu/airflow/logs/request_logs/request_logs.log")
+logging.getLogger().setLevel(logging.DEBUG)
+requests_log = logging.getLogger("requests.packages.urllib3")
+requests_log.setLevel(logging.DEBUG)
+requests_log.propagate = True
+
+BASE_PATH = "/scratch/project_2006633/nlf-harvester/downloads"
+SSH_CONN_ID = "puhti_conn"
+HTTP_CONN_ID = "nlf_http_conn"
+SET_IDS = ["col-681", "col-361"]
+
+default_args = {
+    "owner": "Kielipankki",
+    "start_date": "2022-10-01",
+    "retry_delay": timedelta(seconds=10),
+}
+
+
+def list_dc_identifiers(http_conn_id, set_id):
+    """
+    A hopefully temporary helper function to save DC identifiers to file for
+    later reading while rate limits are an issue.
+    """
+    if not os.path.isfile(
+        f"/home/ubuntu/airflow/plugins/bindings_{set_id.replace(':', '_')}"
+    ):
+        http_conn = BaseHook.get_connection(http_conn_id)
+        api = PMH_API(url=http_conn.host)
+        dc_identifiers = api.dc_identifiers(set_id)
+        with open(
+            f"/home/ubuntu/airflow/plugins/bindings_{set_id.replace(':', '_')}", "w"
+        ) as file:
+            for dc in dc_identifiers:
+                file.write(f"{dc}\n")
+
+
+# Again, temporary until rate limits are no longer an issue
+for set_id in SET_IDS:
+    list_dc_identifiers(HTTP_CONN_ID, set_id)
+
+
+def download_set(
+    dag: DAG, set_id, api, ssh_conn_id, base_path, trigger_rule
+) -> TaskGroup:
+    """
+    TaskGroupFactory for downloading METS and ALTOs for one binding.
+    """
+    with TaskGroup(group_id=f"download_set_{set_id.replace(':', '_')}") as download:
+
+        # Temporary solution while rate limits are an issue:
+        with open(
+            f"/home/ubuntu/airflow/plugins/bindings_{set_id.replace(':', '_')}"
+        ) as file:
+            dc_identifiers = file.read().splitlines()
+
+        # This is how it would preferrably be done:
+        # expand() only accepts lists or dicts, so dc_identifiers need to be
+        # converted to a list. With the current implementation, this would
+        # be an issue for large sets, but we have an upcoming ticket to
+        # download huge sets in batches anyway, which should hopefully fix
+        # this issue.
+        # dc_identifiers = list(api.dc_identifiers(set_id))
+
+        @task(
+            task_id=f"download_binding", task_group=download, trigger_rule=trigger_rule
```

Comment:
This f-stringness looks unnecessary

Reply:
It does indeed, this is leftover from when I was hopeful that we could somehow give a custom name for each mapped task with `expand()`. Turns out we cannot without creating the tasks in a loop.